### PR TITLE
fix: apply filterFunc in NRHttpTracer for non-empty pattern (#42)

### LINKED ISSUE #42: NRHttpTracer: filterFunc not applied when pattern is non-empty
When pattern is non-empty, NRHttpTracer returns newrelic.WrapHandleFunc directly without checking filterFunc. Filtered routes still get traced.

### PR DESCRIPTION
**Pull request by ankurs**
## Summary

When `pattern` was non-empty, `NRHttpTracer` returned `newrelic.WrapHandleFunc` directly and never consulted the configured `filterFunc`. Result: filtered routes (health checks, reflection, anything rejected by the filter) still started a New Relic transaction — exactly what the filter is meant to prevent. The empty-pattern branch already did the check correctly.

Fix: in the non-empty-pattern branch, wrap the handler returned by `WrapHandleFunc` with a filter-aware shell. Filtered paths fall through to the underlying handler; non-filtered paths go through the NR-instrumented handler as before.

Fixes #42.

## Test plan

- [x] New test `TestNRHttpTracer_FilterFuncAppliedForNonEmptyPattern` asserts a filtered path has no NR transaction in request context and a passing path has one. Verified the test fails on the pre-fix code.
- [x] Companion test `TestNRHttpTracer_FilterFuncAppliedForEmptyPattern` guards the other branch.
- [x] `TestNRHttpTracer_NilApp` confirms the nil-app short-circuit still returns the original handler unchanged.
- [x] `go test -race ./...` passes.
- [x] README regenerated via `make doc`.

https://claude.ai/code/session_01EWgytCisKXLiVp5BNhp2xE

## Release Notes

* **New Features**
  * Per-request filtering now determines whether New Relic instrumentation is applied based on request paths.

* **Documentation**
  * Documentation updated with detailed explanation of instrumentation behavior across different routing patterns and conditions.

* **Tests**
  * Comprehensive test coverage added for conditional instrumentation behavior and edge cases.

### CODE REVIEW

File: http.go [hunk 1]
```diff
@@ -57,14 +57,25 @@ func DoHTTPtoGRPC(ctx context.Context, svr any, handler func(ctx context.Context
...
 	}
 	if pattern != "" {
-		return newrelic.WrapHandleFunc(app, pattern, h)
+		wrappedPattern, wrappedHandler := newrelic.WrapHandleFunc(app, pattern, h)
+		return wrappedPattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if defaultConfig.filterFunc(context.Background(), r.URL.Path) {
+				wrappedHandler(w, r)
+				return
+			}
+			h(w, r)
```
[the remark this answers is not in the document]
Comment (ankurs) [reply http.go:77]: Addressed in caa1187: `NRHttpTracer` now passes `r.Context()` to `filterFunc` in both the non-empty and empty pattern branches, so request-scoped values reach the filter.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: interceptors_test.go [hunk 2]
```diff
@@ -2197,3 +2200,106 @@ func TestDefaultStreamInterceptors_UserInterceptorsOutermost(t *testing.T) {
...
+	h := http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(200) })
+
+	p, got := NRHttpTracer("/route", h)
+	if p != "/route" {
+		t.Errorf("pattern should be unchanged when NR app is nil; got %q", p)
+	}
+	// With no NR app, the function must return the original handler reference.
+	if fmt.Sprintf("%p", got) != fmt.Sprintf("%p", h) {
+		t.Error("expected original handler to be returned unchanged when NR app is nil")
+	}
```
[the remark this answers is not in the document]
Comment (ankurs) [reply interceptors_test.go:2237]: Addressed in caa1187: replaced `fmt.Sprintf("%p", ...)` with `reflect.ValueOf(got).Pointer() == reflect.ValueOf(h).Pointer()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: http.go [hunk 3]
```diff
@@ -57,18 +57,29 @@ func DoHTTPtoGRPC(ctx context.Context, svr any, handler func(ctx context.Context
...
+		return wrappedPattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if defaultConfig.filterFunc(r.Context(), r.URL.Path) {
+				wrappedHandler(w, r)
+				return
+			}
+			h(w, r)
+		})
 	}
 	return pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		// filter functions we do not need
```
[the remark this answers is not in the document]
Comment (ankurs) [reply http.go]: Addressed in be01ef0: dropped the stale `// filter functions we do not need` comment. The `defaultConfig.filterFunc(...)` call is self-documenting; no replacement comment needed.

---
_Generated by [Claude Code](https://claude.ai/code)_
